### PR TITLE
Enable obscore for casa ms

### PR DESCRIPTION
**Pull request by Lukas113**
- Implemented CASA MS metadata & data reading utility in `karabo.data.casa.py`
- Implemented `ObsCoreMeta.from_visibility` CASA MS support
- Bugfix: `ObsCoreMeta.from_visibility` for OSKAR-vis `t_exptime` in sec, not mjd
- Adapted `SRCNet_rucio_meta.py` script to use CASA MS instead of OSKAR-vis.

### CODE REVIEW

File: karabo/data/casa.py [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+    return cols
+
+
+def _create_table(
+    table: casa_table,
+    classtype: Type[_TDataClass],
+    *,
+    subtable_id: Optional[int] = None,
+) -> _TDataClass:
+    """Creates an instance of of `classtype` from `table` values.
```
Comment (mpluess) [inline karabo/data/casa.py:66]: one of is enough ;)

File: karabo/data/casa.py [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+    This function assumes that ALL fields of `classtype` are available in `table` in
```
Comment (mpluess) [inline karabo/data/casa.py:68]: Formulation could be clearer. "Assumes" all fields are available in table but actually handles the case where they're not. Maybe explicitly say it throws a warning if not all fields are available? And that None is assigned for missing fields.

File: karabo/data/obscore.py [hunk 3]
```diff
@@ -421,7 +467,7 @@ def check_ObsCoreMeta(
     @classmethod
     def from_visibility(
         cls,
-        vis: Visibility,
+        vis: Visibility,  # not sure if also MS-path should be allowed
```
Comment (mpluess) [inline karabo/data/obscore.py:470]: Dont' think MS path should be allowed the way we're designing it now, supporting multiple formats

File: karabo/data/obscore.py [hunk 4]
```diff
@@ -433,26 +479,94 @@ def from_visibility(
...
             `ObsCoreMeta` instance.
         """
         ocm = cls(dataproduct_type="visibility")
-        # only .vis supported atm, `OSKAR` can't read .ms with multiple spectral windows
-        # for multiple format-support, think about restructuring the entire function,
-        # not just another elif block!
-        vis_path = vis.vis_path
-        if os.path.exists(vis_path):
-            header, _ = VisHeader.read(vis_path)
+        vis_inode = vis.visibility_path
```
Comment (mpluess) [inline karabo/data/obscore.py:500]: .visibility_path -> .path

File: karabo/examples/SRCNet_rucio_meta.py [hunk 5]
```diff
@@ -1,119 +1,157 @@
...
-    name = os.path.split(dp_path)[-1]
-    rm = RucioMeta(
+    vis_ocm = ObsCoreMeta.from_visibility(
+        vis=vis,
+        calibrated=False,
+    )
+    vis_rm = RucioMeta(
         namespace="testing",  # needs to be specified by Rucio service
-        name=name,
+        name=os.path.split(vis.visibility_path)[-1],  # remove path-infos for `name`
```
Comment (mpluess) [inline karabo/examples/SRCNet_rucio_meta.py:75]: .visibility_path -> .path

File: karabo/examples/SRCNet_rucio_meta.py [hunk 6]
```diff
@@ -1,119 +1,157 @@
...
+        name=vis_rm.name,
+    )
+    vis_ocm.obs_publisher_did = obs_publisher_did
+
+    # fill/correct other fields of `ObsCoreMeta` here!
+    # #####START#####
+    # HERE
+    # #####END#######
+
+    vis_path_meta = RucioMeta.get_meta_fname(fname=vis.visibility_path)
```
Comment (mpluess) [inline karabo/examples/SRCNet_rucio_meta.py:97]: .visibility_path -> .path

File: karabo/examples/SRCNet_rucio_meta.py [hunk 7]
```diff
@@ -1,119 +1,157 @@
...
+    print(f"Imaging: {imaging_npixel=}, {imaging_cellsize=} ...", flush=True)
+    restored_path = os.path.join(FileHandler.stm(), "my-unique-image-fname.fits")
+    restored = WscleanImageCleaner(
+        WscleanImageCleanerConfig(
+            imaging_npixel=imaging_npixel,
+            imaging_cellsize=imaging_cellsize,
+            niter=5000,  # 10 times less than default
+        )
+    ).create_cleaned_image(  # currently, wsclean needs casa .ms, which is also created
+        ms_file_path=vis.visibility_path,
```
Comment (mpluess) [inline karabo/examples/SRCNet_rucio_meta.py:121]: .visibility_path -> .path

File: karabo/data/obscore.py (hunk 4 above)
Comment (mpluess) [reply karabo/data/obscore.py:500]: In general: maybe merge my branch again, do another run to make sure it stills works after the latest changes

File: karabo/data/obscore.py [hunk 8]
```diff
@@ -433,26 +479,94 @@ def from_visibility(
...
+        if not os.path.exists(vis_inode):
+            err_msg = f"{vis_inode} doesn't exist!"
+            raise FileNotFoundError(err_msg)
+        c = const.c.value
+        if vis.visibility_format == "MS":
```
Comment (mpluess) [inline karabo/data/obscore.py:505]: .visibility_format -> .format

File: karabo/data/obscore.py [hunk 9]
```diff
@@ -433,26 +479,94 @@ def from_visibility(
...
+            )
+            b = float(np.max(ms_meta.antenna.baseline_dists()))
+            ocm.s_resolution = Telescope.ang_res(
+                freq=float(np.max(spectral_window.ref_frequency)),
+                b=b,
+            )
+            ocm.pol_xel = MSPolarizationTable.nrows(ms_path=vis_inode)
+            corr_types = np.unique(ms_meta.polarization.corr_type.ravel()).tolist()
+            ocm.set_pol_states(pol_states=corr_types)
+        elif vis.visibility_format == "OSKAR_VIS":
```
Comment (mpluess) [inline karabo/data/obscore.py:568]: .visibility_format -> .format

File: karabo/data/obscore.py [hunk 10]
```diff
@@ -483,33 +596,33 @@ def from_visibility(
...
+            if tel is not None and (tel_name := tel.name) is not None:
+                ocm.instrument_name = tel_name
+            if tel is not None and obs is not None:
+                freq_inc_hz = obs.frequency_increment_hz
+                min_freq_hz = obs.start_frequency_hz - freq_inc_hz / 2
+                end_freq_hz = min_freq_hz + freq_inc_hz * obs.number_of_channels
+                b = float(tel.max_baseline())
+                ocm.s_resolution = tel.ang_res(freq=end_freq_hz, b=b)
+        else:
+            assert_never(vis.visibility_format)
```
Comment (mpluess) [inline karabo/data/obscore.py:608]: .visibility_format -> .format

File: karabo/data/casa.py (hunk 2 above)
Comment (Lukas113) [reply karabo/data/casa.py:68]: You're right. This was an assumption in an older commit.

**Review by mpluess — APPROVED** [review]
LGTM (merge conflict needs to be solved though)

**Pull request merged**